### PR TITLE
[WIP] particle filtering

### PR DESCRIPTION
**Pull request by mneedham**

### CODE REVIEW

**Comment by mneedham** [comment]
cc / @kuzeko, @velgias, @DenisGallo

File: alpha/alpha-proc/src/main/java/org/neo4j/graphalgo/centrality/ParticleFilteringProc.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+public class ParticleFilteringProc {
```
Comment (kuzeko) [inline alpha/alpha-proc/src/main/java/org/neo4j/graphalgo/centrality/ParticleFilteringProc.java:40]: The Javadoc of the class should contain the full reference:

> Gallo, Denis; Lissandrini, Matteo; and Velegrakis, Yannis.
“Personalized Page Rank on Knowledge Graphs: Particle Filtering is all you need!.”
Proceedings of the 23th International Conference on Extending Database Technology, EDBT 2020 (169-180). 
http://dx.doi.org/10.5441/002/edbt.2020.54

File: alpha/alpha-proc/src/main/java/org/neo4j/graphalgo/centrality/ParticleFilteringProc.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            for (Long node : p.keySet()) {
+                double particles = p.get(node) * dampingFactor;
+                Node startingNode = transaction.internalTransaction().getNodeById(node);
+
+                List<Long> neighbours = StreamSupport
+                        .stream(getRelationships(startingNode, relationshipType, direction).spliterator(), false)
+                        .map(relationship -> relationship.getOtherNode(startingNode).getId()).collect(Collectors.toList());
+                int neighboursCount = neighbours.size();
+
+                if(shuffleNeighbors) {
```
Comment (kuzeko) [inline alpha/alpha-proc/src/main/java/org/neo4j/graphalgo/centrality/ParticleFilteringProc.java:90]: This version of the code is the `unlabelled` version. It is not recommended to use the unlabelled version in large graphs. The `labelled` version should make use of a weighting function on edges.

File: doc/asciidoc/algorithms/alpha/alpha-particle-filtering.adoc [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+* <<algorithms-particle-filtering-context, History and explanation>>
+* <<algorithms-particle-filtering-usecase, Use-cases - when to use the Particle Filtering algorithm>>
+* <<algorithms-particle-filtering-syntax, Syntax>>
+* <<algorithms-particle-filtering-examples, Examples>>
+
+[[algorithms-particle-filtering-context]]
+== History and explanation
+
+Particle Filtering was proposed by Denis Gallo, Matteo Lissandrini, and Yannis Velegrakis in their 2020 paper https://people.cs.aau.dk/~matteo/pdf/EDBT20-particle-filtering.pdf[Personalized Page Rank on Knowledge Graphs: Particle Filtering is all you need!^].
```
Comment (kuzeko) [inline doc/asciidoc/algorithms/alpha/alpha-particle-filtering.adoc:23]: The full reference for the core Particle Filtering Idea is:

> Lao, Ni, and William W. Cohen. "Fast query execution for retrieval models based on path-constrained random walks." Proceedings of the 16th ACM SIGKDD international conference on Knowledge discovery and data mining. 2010.
> https://doi.org/10.1145/1835804.1835916

While full reference the specific algorithm refers to this

> Gallo, Denis; Lissandrini, Matteo; and Velegrakis, Yannis.
“Personalized Page Rank on Knowledge Graphs: Particle Filtering is all you need!.”
Proceedings of the 23th International Conference on Extending Database Technology, EDBT 2020 (169-180). 
> http://dx.doi.org/10.5441/002/edbt.2020.54

File: doc/asciidoc/algorithms/alpha/alpha-particle-filtering.adoc [hunk 4]
```diff
@@ -0,0 +1,99 @@
+[[algorithms-particle-filtering-centrality]]
+[.alpha]
+= Particle Filtering
+
+[abstract]
+--
+This section describes the Particle Filtering algorithm in the Neo4j Graph Data Science library.
+--
+
+Particle filtering measures the proximity between source nodes and other nodes in the graph.
```
Comment (kuzeko) [inline doc/asciidoc/algorithms/alpha/alpha-particle-filtering.adoc:10]: Particle filtering methods produce an approximation of the Personalized Page Rank (PPR) score with sufficient precision while exploiting what graph databases architectures are already optimized for: navigating local connections.

Personalized Page Rank (PPR) is a particularly important task to support search and exploration within graphs. 
PPR computes the proximity between query nodes and other *local* nodes in the graph.
In practice, given a small set of query entities, PPR returns a ranked list of other relevant entities based on a computed random-walk proximity to the query nodes.
This is used, among others, for entity exploration, query expansion, and product recommendation.

File: doc/asciidoc/algorithms/alpha/alpha-particle-filtering.adoc [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+[source, cypher, role=setup-query]
+----
+include::../page-rank/sample-graph.cypher[tag=setup]
+----
+
+The following examples show how to run Particle Filtering centered around 'Site A'.
+
+[role=query-example]
+--
+.The following will run the algorithm and stream results:
```
Comment (kuzeko) [inline doc/asciidoc/algorithms/alpha/alpha-particle-filtering.adoc:73]: It is important to explain the meaning of the parameters and how to select them.
In particular, the number of particles w.r.t. the minimum threshold.

Comment (mneedham) [reply doc/asciidoc/algorithms/alpha/alpha-particle-filtering.adoc:81]: > * The documentation is missing the explanation for the parameters, these are fundamental to obtain good results (p.s. I can provide more text for the documentation about these)

Have updated the citations and added more documentation. Could you provide some more text for documenting the parameters? 

I'll add the labelled version of the algorithm as well. I just want to check that I understand how the weights are computed here: https://github.com/DenisGallo/Neo4j-ParticleFiltering/blob/master/src/main/java/pfiltering/WeightingGraphProc.java#L30

* Get total relationship count
* Get the count of each relationship type
* Store a score on each relationship of: `-1 * log(1.0*relTypeCount/totalRelCount)`

If that's the correct formula, I don't think we need to compute that in a separate procedure. Neo has internal stats stores that count the number of total relationships & the number of each type of relationship, so we can retrieve those inside the particleFiltering procedure call at low cost.

Comment (kuzeko) [reply doc/asciidoc/algorithms/alpha/alpha-particle-filtering.adoc:81]: **Parameters explanations:**
`numberParticles`:   controls the granularity of the computation, and consequently the precision of the final result. An ideal initial number of particles is some multiple of the degree of the source query  nodes.

`minThreshold`: minimum should be `1/numberParticles` . This determines the minimum amount of PPR score for  a node to be returned as result.  if set to `1/numberParticles` it should return all nodes visited. Higher value will return only nodes with some non trivial score.


`dampingFactor` : this controls how far from the source nodes the PPR algorithm should navigate. `0.85` usually provides a good balance. For a very sparse and large graph (in terms of diameter) values around `0.9-0.95` can be also used.

**Comment by kuzeko** [comment]
Hi @mneedham  please let me know if you need further support on this

**Pull request closed without merging**